Item Pool Refactor

Cuphat opened the pull request:
`ItemPool.py` was an absolute tire fire. Hopefully it is less of one now. `DungeonList.py` got caught up in the crossfire, too.

This seems to work and tests pass. Regardless, this thing will need some testing.

Cuphat committed:
Refactor Item Pool: Gold Skulltulas

Cuphat committed:
Refactor Item Pool: Shops/Milk

Cuphat committed:
Refactor Item Pool: Maps/Compasses/Keys

Cuphat committed:
Refactor Item Pool: Items in Dungeons

Shopsanity loses 3 "Rupees (20)" and 3 "Rupees (50)" in exchange for 6 "Rupees (5)" and the sanity of this file.

Cuphat committed:
Refactor Item Pool: Business Scrubs

Cuphat committed:
Refactor Item Pool: Always Placed/Bottles/Conditional Shuffles

Cuphat committed:
Refactor Item Pool: Dungeons/Thieves' Hideout

Cuphat committed:
Refactor Item Pool: Pending Junk Pool

Cuphat committed:
Refactor Item Pool: Full Item Pool

Now uses `location.vanilla_item` for all applicable places in the item pool.

Cuphat commented:
Already see more shit to remove or consolidate, honestly.

Cuphat committed:
Refactor Item Pool: Final Cleanup

This should hopefully be the last time I touch `get_pool_core` in this particular refactor.

Zannick commented on Dungeon.py:
```diff
@@ -1,6 +1,6 @@
 
-    def __init__(self, world, name, hint, font_color, boss_key, small_keys, dungeon_items):
+    def __init__(self, world, name, hint, font_color, boss_key=None, small_keys=None, dungeon_items=None):
         def to_array(obj):
```
can now be removed

Cuphat replied:
It's still used in `copy`.

Zannick replied:
uh... what? How can it be? `to_array` is a local to `__init__`, and not used anymore?

Cuphat replied:
Resolved.

Cuphat replied:
The review showing the lines above confused the hell out of me as to what you were referring to.

Zannick commented on ItemPool.py:
```diff
@@ -813,368 +323,226 @@ def collect_heart_container(world, pool):
+
+        # Gold Skulltula Tokens
+        elif location.vanilla_item == 'Gold Skulltula Token':
+            shuffle_item = world.settings.tokensanity == 'all' \
```
nit: wrap in parens instead of using `\\`

Cuphat replied:
Resolved.

cjohnson57 commented on ItemPool.py:
```diff
@@ -813,524 +323,253 @@ def collect_heart_container(world, pool):
+                    item = get_junk_item()[0]
+                    shuffle_item = True
+            # Any other item in a dungeon.
+            elif location.type in ["Chest", "NPC", "Song", "Collectable", "Cutscene", "BossHeart"]:
```
"Scrub" should be added to this list, right?

Cuphat replied:
Scrubs are handled earlier.

cjohnson57 commented on Rules.py:
```diff
@@ -44,7 +44,7 @@ def set_rules(world):
             else:
                 add_item_rule(location, lambda location, item: item.type == 'Shop' and item.world.id == location.world.id)
-        elif 'Deku Scrub' in location.name:
+        elif location.type in ['Scrub', 'GrottoNPC']:
```
While we're at it, feel like we may as well rename GrottoNPC to GrottoScrub, since it's not like there are any other kinds of NPCs in grottos.

Cuphat replied:
Yeah, I thought about it and I agree. Done.

cjohnson57 reviewed (commented):
Looks good for the most part

Cuphat commented on ItemPool.py:
```diff
@@ -813,524 +323,253 @@ def collect_heart_container(world, pool):
+        if world.settings.shuffle_gerudo_card:
+            pending_junk_pool.append('Gerudo Membership Card')
+        if world.settings.shuffle_smallkeys in ['any_dungeon', 'overworld', 'keysanity']:
+            for dungeon in ['Forest Temple', 'Fire Temple', 'Water Temple', 'Shadow Temple', 'Spirit Temple', 'Ganons Castle']:
```
Indeed. Good catch.

Cuphat committed:
Rename "GrottoNPC" to "GrottoScrub".

Cuphat committed:
Refactor Item Pool: Final Final Cleanup

Fixes a Thieves' Hideout bug on Plentiful Item Pools + Key Ring. Merges the trade_item lists into one OrderedDict. Ship it.

Cuphat committed:
Define Junk Items in ItemList.py

Junk items are now defined by adding a 'junk' entry to the item's `special` dictionary. The value given is the weight that junk item gets in the junk_pool. If the weight is 0, the item is not added to the junk_pool but still is added to remove_junk_items. If the weight is negative, the item is still classified as junk but will not be in either junk_pool or remove_junk_items.

Cuphat committed:
Replace item_table usages with ItemInfo.items

ItemInfo.items is a much more usable version of item_table.

Cuphat committed:
Simplify Key Ring handling in Plandomizer.py

Cuphat committed:
ItemPool.py: Use ItemInfo for Bottles, Songs, Dungeon Rewards

Instead of explicitly defining these lists, create them from the ItemInfo data.

cjohnson57 reviewed (approved):
Everything seems good to me, these are some great changes

fenhl reviewed (approved):
Thank you for taking the time to go through and clean up this file. This is pretty much exactly what I was working on when I sent #1242, except your version actually seems to work (and without major changes to Fill.py, too).

Cuphat committed:
ItemPool.py: Remove unused Piece of Heart/Heart Container functions.

These functions were added for Starting Hearts, but they were rendered obsolete and unused in commit c6eeaf815d79f4af12227d21c6b99750e37fc87a.
